File: analysis_engine/services/investment.py
"""Services for single-company investment analysis."""

from __future__ import annotations

from typing import Optional

import pandas as pd

from analysis_engine.calculations import calculate_event_return, calculate_max_drawdown
from analysis_engine.data_access import load_events, load_prices
from analysis_engine.services.company import build_company_snapshot
from analysis_engine.services.comparison import build_investment_comparison
from analysis_engine.services.serialization import to_jsonable
from analysis_engine.utils import ceo_events, row_on_or_after, row_on_or_before


def build_investment_analysis(
    ticker: str,
    benchmark: str = "SPY",
    initial_amount: float = 10_000,
    years: Optional[float] = None,
    start_date: Optional[str] = None,
    end_date: Optional[str] = None,
) -> dict:
    """Build a JSON-ready end-to-end company investment analysis."""
    ticker = ticker.upper()
    comparison = build_investment_comparison(
        ticker,
        benchmark,
        initial_amount,
        years=years,
        start_date=start_date,
        end_date=end_date,
    )

    prices = load_prices(ticker)
    events = load_events(ticker)
    resolved_start = pd.Timestamp(comparison["resolved_dates"]["start_date"])
    resolved_end = pd.Timestamp(comparison["resolved_dates"]["end_date"])
    path = prices[(prices["date"] >= resolved_start) & (prices["date"] <= resolved_end)].copy()
    start_row = row_on_or_after(path, resolved_start)
    path["wealth"] = initial_amount * path["adj_close"] / start_row["adj_close"]

    drawdown = calculate_max_drawdown(path["wealth"].tolist())
    ceo_event_table = ceo_events(events, resolved_start, resolved_end)

    return to_jsonable(
        {
            "input": {
                "ticker": ticker,
                "benchmark": benchmark.upper(),
                "initial_amount": initial_amount,
                "years": years,
                "start_date": start_date,
                "end_date": end_date,
            },
            "comparison": comparison,
            "company_snapshot": build_company_snapshot(ticker),
            "risk": {
                "max_drawdown": drawdown,
                "peak_date": path.iloc[drawdown.peak_index]["date"],
                "trough_date": path.iloc[drawdown.trough_index]["date"],
                "recovery_date": (
                    None
                    if drawdown.recovery_index is None
                    else path.iloc[drawdown.recovery_index]["date"]
                ),
            },
            "events": {
                "ceo_events": ceo_event_table.to_dict(orient="records"),
                "ceo_event_reactions": _build_ceo_event_reactions(
                    prices,
                    ceo_event_table,
                    periods_before=60,
                    periods_after=120,
                ),
            },
        }
    )


def _build_ceo_event_reactions(
    prices: pd.DataFrame,
    ceo_event_table: pd.DataFrame,
    periods_before: int,
    periods_after: int,
) -> list[dict]:
    reactions = []
    price_values = prices["adj_close"].tolist()
    for _, event in ceo_event_table.drop_duplicates(subset=["date"]).iterrows():
        event_row = row_on_or_after(prices, event["date"])
        reaction = calculate_event_return(
            price_values,
            int(event_row.name),
            periods_before=periods_before,
            periods_after=periods_after,
        )
        reactions.append(
            {
                "date": event["date"],
                "executive_name": event["executive_name"],
                "action": event["action"],
                "reaction": reaction,
            }
        )
    return reactions

File: lib/analysis_engine/services/investment.js
import { calculate_event_return } from "../calculations/events";
import { calculate_max_drawdown } from "../calculations/drawdowns";
import { load_events, load_prices, normalize_ticker } from "../data_access/public_data";
import { ceo_events } from "../utils/events";
import { row_on_or_after } from "../utils/dates";
import { build_company_snapshot } from "./company";
import { build_investment_comparison } from "./comparison";

export async function build_investment_analysis({
  ticker,
  benchmark = "SPY",
  initial_amount = 10_000,
  years = null,
  start_date = null,
  end_date = null,
}) {
  const normalized = normalize_ticker(ticker);
  const [comparison, prices, events, company_snapshot] = await Promise.all([
    build_investment_comparison({
      first_ticker: normalized,
      second_ticker: benchmark,
      initial_amount,
      years,
      start_date,
      end_date,
    }),
    load_prices(normalized),
    load_events(normalized),
    build_company_snapshot(normalized),
  ]);
  const resolved_start = comparison.resolved_dates.start_date;
  const resolved_end = comparison.resolved_dates.end_date;
  const start_row = row_on_or_after(prices, resolved_start);
  const path = prices
    .filter((row) => row.date >= resolved_start && row.date <= resolved_end)
    .map((row) => ({ ...row, wealth: (initial_amount * row.adj_close) / start_row.adj_close }));
  const drawdown = calculate_max_drawdown(path.map((row) => row.wealth));
  const ceo_event_rows = ceo_events(events, resolved_start, resolved_end);

  return {
    input: { ticker: normalized, benchmark: normalize_ticker(benchmark), initial_amount, years, start_date, end_date },
    comparison,
    company_snapshot,
    risk: {
      max_drawdown: drawdown,
      peak_date: path[drawdown.peak_index]?.date,
      trough_date: path[drawdown.trough_index]?.date,
      recovery_date: drawdown.recovery_index == null ? null : path[drawdown.recovery_index]?.date,
    },
    events: {
      ceo_events: ceo_event_rows,
      ceo_event_reactions: build_ceo_event_reactions(prices, ceo_event_rows, 60, 120),
    },
  };
}

function build_ceo_event_reactions(prices, ceo_event_rows, periods_before, periods_after) {
  const price_values = prices.map((row) => row.adj_close);
  return ceo_event_rows.map((event) => {
    const event_index = prices.findIndex((row) => row.date >= event.date);
    return {
      date: event.date,
      executive_name: event.executive_name,
      action: event.action,
      reaction: calculate_event_return(price_values, event_index, periods_before, periods_after),
    };
  });
}
